libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/ptx.h

```diff
@@ -805,6 +805,1641 @@ _CCCL_DEVICE static inline void cp_async_bulk(
+// .dst       = { .global }
+// .src       = { .shared::cta }
+// .type      = { .f64 }
+// .op        = { .add }
+template <typename=void>
+__device__ static inline void cp_reduce_async_bulk(
+  cuda::ptx::space_global_t,
+  cuda::ptx::space_shared_t,
+  cuda::ptx::op_add_t,
+  double* dstMem,
+  const double* srcMem,
+  uint32_t size);
+*/
+#if __cccl_ptx_isa >= 800
+extern "C" _CCCL_DEVICE void __cuda_ptx_cp_reduce_async_bulk_is_not_supported_before_SM_90__();
+template <typename=void>
+_CCCL_DEVICE static inline void cp_reduce_async_bulk(
+  space_global_t,
+  space_shared_t,
+  op_add_t,
+  double* __dstMem,
+  const double* __srcMem,
+  _CUDA_VSTD::uint32_t __size)
+{
+  // __space == space_global (due to parameter type constraint)
+  // __space == space_shared (due to parameter type constraint)
+  // __type == type_f64 (due to parameter type constraint)
+  // __op == op_add (due to parameter type constraint)
+  NV_IF_ELSE_TARGET(NV_PROVIDES_SM_90,(
+    asm (
+      "cp.reduce.async.bulk.global.shared::cta.bulk_group.add.f64  [%0], [%1], %2; // 4."
+      :
+      : "l"(__as_ptr_gmem(__dstMem)),
+        "r"(__as_ptr_smem(__srcMem)),
+        "r"(__size)
+      : "memory"
+    );
+  ),(
+    // Unsupported architectures will have a linker error with a semi-decent error message
+    __cuda_ptx_cp_reduce_async_bulk_is_not_supported_before_SM_90__();
+  ));
+}
+#endif // __cccl_ptx_isa >= 800
+
+/*
+// cp.reduce.async.bulk.dst.src.bulk_group.op.noftz.type  [dstMem], [srcMem], size; // 5. PTX ISA 80, SM_90
+// .dst       = { .global }
+// .src       = { .shared::cta }
+// .type      = { .f16 }
+// .op        = { .add }
+template <typename=void>
+__device__ static inline void cp_reduce_async_bulk(
+  cuda::ptx::space_global_t,
+  cuda::ptx::space_shared_t,
+  cuda::ptx::op_add_t,
+  __half* dstMem,
+  const __half* srcMem,
+  uint32_t size);
+*/
+#if __cccl_ptx_isa >= 800
+extern "C" _CCCL_DEVICE void __cuda_ptx_cp_reduce_async_bulk_is_not_supported_before_SM_90__();
+template <typename=void>
+_CCCL_DEVICE static inline void cp_reduce_async_bulk(
+  space_global_t,
+  space_shared_t,
+  op_add_t,
+  __half* __dstMem,
+  const __half* __srcMem,
+  _CUDA_VSTD::uint32_t __size)
+{
+  // __space == space_global (due to parameter type constraint)
+  // __space == space_shared (due to parameter type constraint)
+  // __type == type_f16 (due to parameter type constraint)
+  // __op == op_add (due to parameter type constraint)
+  NV_IF_ELSE_TARGET(NV_PROVIDES_SM_90,(
+    asm (
+      "cp.reduce.async.bulk.global.shared::cta.bulk_group.add.noftz.f16  [%0], [%1], %2; // 5."
+      :
+      : "l"(__as_ptr_gmem(__dstMem)),
+        "r"(__as_ptr_smem(__srcMem)),
+        "r"(__size)
+      : "memory"
+    );
+  ),(
+    // Unsupported architectures will have a linker error with a semi-decent error message
+    __cuda_ptx_cp_reduce_async_bulk_is_not_supported_before_SM_90__();
+  ));
+}
+#endif // __cccl_ptx_isa >= 800
```

Comment:
Those need to additionally be guarded similar to what we do [here ](https://github.com/NVIDIA/cccl/pull/1140/files#diff-b83299cb0c96e7eace7ddd92594dfc7d3beaa7e80c62cde9eb9629fc0b324b82)

Reply:
Good catch. Do you mean the one overload should be guarded by `_LIBCUDACXX_HAS_NVBF16` and the other by `_LIBCUDACXX_HAS_NVF16`?

Reply:
Yeah we need to guard all `__half` behind `_LIBCUDACXX_HAS_NVF16` and all `__nv_bfloat16` behind `_LIBCUDACXX_HAS_NVBF16 `